fix(windows-path): preserve PATH order; include core env vars

cpjet64 opened the pull request:
# Preserve PATH precedence & fix Windows MCP env propagation

## Problem & intent

Preserve user PATH precedence and reduce Windows setup friction for MCP servers by avoiding PATH reordering and ensuring Windows child processes receive essential env vars.

- Addresses: #4180 #5225 #2945 #3245 #3385 #2892 #3310 #3457 #4370  
- Supersedes: #4182, #3866, #3828 (overlapping/inferior once this merges)  
- Notes: #2626 / #2646 are the original PATH-mutation sources being corrected.

---

## Before / After

**Before**  
- PATH was **prepended** with an `apply_patch` helper dir (Rust + Node wrapper), reordering tools and breaking virtualenvs/shims on macOS/Linux.  
- On Windows, MCP servers missed core env vars and often failed to start without explicit per-server env blocks.

**After**  
- Helper dir is **appended** to PATH (preserves user/tool precedence).  
- Windows MCP child env now includes common core variables and mirrors `PATH` → `Path`, so typical CLIs/plugins work **without** per-server env blocks.

---

## Scope of change

### `codex-rs/arg0/src/lib.rs`
- Append temp/helper dir to `PATH` instead of prepending.

### `codex-cli/bin/codex.js`
- Mirror the same append behavior for the Node wrapper.

### `codex-rs/rmcp-client/src/utils.rs`
- Expand Windows `DEFAULT_ENV_VARS` (e.g., `COMSPEC`, `SYSTEMROOT`, `PROGRAMFILES*`, `APPDATA`, etc.).  
- Mirror `PATH` → `Path` for Windows child processes.  
- Small unit test; conditional `mut` + `clippy` cleanup.

---

## Security effects

No broadened privileges. Only environment propagation for well-known Windows keys on stdio MCP child processes. No sandbox policy changes and no network additions.

---

## Testing evidence

**Static**  
- `cargo fmt`  
- `cargo clippy -p codex-arg0 -D warnings` → **clean**  
- `cargo clippy -p codex-rmcp-client -D warnings` → **clean**  
- `cargo test -p codex-rmcp-client` → **13 passed**

**Manual**  
- Local verification on Windows PowerShell 5/7 and WSL (no `unused_mut` warnings on non-Windows targets).

---

## Checklist

- [x] Append (not prepend) helper dir to PATH in Rust and Node wrappers  
- [x] Windows MCP child inherits core env vars; `PATH` mirrored to `Path`  
- [x] `cargo fmt` / `clippy` clean across touched crates  
- [x] Unit tests updated/passing where applicable  
- [x] Cross-platform behavior preserved (macOS/Linux PATH precedence intact)

cpjet64 committed:
[fix][windows-path]: preserve PATH order; include core env vars

Context:
* Address open issues related to PATH/PATHEXT and Windows env propagation.
* Prepending to PATH changed users' tool precedence (e.g., virtualenv). MCP servers on Windows missed critical env vars (COMSPEC, SYSTEMROOT, etc.).

Change:
* Append apply_patch temp dir to PATH instead of prepending (arg0, Node wrapper).
* Expand Windows DEFAULT_ENV_VARS for MCP client to include COMSPEC, SYSTEMROOT, PROGRAMFILES, APPDATA, etc.

Risk:
* Low: only affects PATH ordering (preserved) and adds allowed env passthrough. No privilege broadening.

Tests:
* Existing rmcp-client unit tests pass.

Verification:
* Ran cargo fmt.
* Built and ran tests: `cargo test -p codex-rmcp-client` (all passed). `cargo check -p codex-arg0` OK.

cpjet64 committed:
[fix][windows-env]: add Path alias and common shell vars

Context:
* Some Windows programs (incl. MCP servers/plugins) read canonical `Path` key.
* Users had to manually set PWSH/POWERSHELL; include when present.

Change:
* In rmcp-client env builder, ensure `Path` mirrors `PATH` on Windows.
* Add PROGRAMW6432, POWERSHELL, PWSH to the default pass-through list.

Risk:
* Low; only adds env keys when available. No privilege expansion.

Tests:
* Added Windows unit test verifying PATH/Path aliasing.

Verification:
* cargo fmt; cargo clippy -p codex-rmcp-client; cargo test -p codex-rmcp-client (all ok).

cpjet64 committed:
[chore][lint]: avoid unused_mut on non-Windows; drop redundant clone

- Conditionally declare `mut` only on Windows in env builder.
- Remove unnecessary `clone()` flagged by clippy.

Validated with: cargo clippy -p codex-rmcp-client -D warnings.

cpjet64 committed:
[fix][windows-path]: mirror Path <-> PATH consistently

- On Windows, if only `Path` is provided (extra env/whitelist), ensure `PATH`
  is set to the same value (and vice versa). Avoids HashMap ordering races.

Validated: clippy clean, rmcp-client tests pass.

cpjet64 commented on codex-rs/rmcp-client/src/utils.rs:
```diff
@@ -76,13 +76,37 @@ pub(crate) fn create_env_for_mcp_server(
+            map.entry("Path".to_string()).or_insert(path_val);
+        } else if let Ok(system_path) = env::var("PATH") {
+            map.insert("PATH".to_string(), system_path.clone());
+            map.entry("Path".to_string()).or_insert(system_path);
```
Good point. I implemented the suggested fix to mirror PATH both directions on Windows so overrides via Path don’t get lost. Logic now sets both PATH and Path to the same value, preferring (in order): explicit PATH, then Path, then the system PATH. Good bot!

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

dylan-hurd-oai reviewed (commented):
Commenting to say that this change feels reasonable to me, but am going to test this before approving

dylan-hurd-oai commented on codex-rs/rmcp-client/src/utils.rs:
```diff
@@ -171,13 +196,33 @@ pub(crate) const DEFAULT_ENV_VARS: &[&str] = &[
     "TMP",
+    // Common shells/pwsh hints
+    "POWERSHELL",
+    "PWSH",
```
@cpjet64 if we separate these changes, i'm happy to stamp the `DEFAULT_ENV_VARS` change and merge it

cpjet64 replied:
@dylan-hurd-oai Understood and I agree the env var change is definitely more low risk because I didn't consider the apply-patch impact in that manner. I'll split them now and get you a new commit in a few minutes! Later I'll dig more into apply-patch to make sure before I push a new PR for it. Thanks for that!

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex

cpjet64 committed:
[fix][windows-env]: scope PR 5579 to DEFAULT_ENV_VARS only\n\nContext:\n* Reviewer requested separating concerns; keep env passthrough low-risk changes in this PR.\n\nChange:\n* Revert PATH precedence changes in arg0 and Node wrapper (restore prepend).\n* Remove PATH<->Path aliasing logic and test from rmcp-client; retain expanded DEFAULT_ENV_VARS.\n\nRisk:\n* None beyond reverting to upstream behavior for PATH order and aliasing; env list expansion preserved.\n\nTests:\n* Ran just fmt; just fix -p codex-arg0, codex-rmcp-client.\n* cargo test -p codex-rmcp-client -> all tests pass.\n\nVerification:\n* Diff for PR now limited to DEFAULT_ENV_VARS in rmcp-client.

cpjet64 commented:
Update: Split the changes as requested.

- Kept this PR focused on Windows env passthrough only: expanded `DEFAULT_ENV_VARS` in `codex-rs/rmcp-client/src/utils.rs`. Removed the `PATH`/`Path` aliasing and reverted PATH ordering changes.
- Moved PATH precedence changes to a new PR: #5898 (append helper dir to PATH in arg0 + Node wrapper).

This should make #5579 low‑risk and ready for review/merge. If you’d like the PATH<->Path mirroring pursued as a separate follow‑up as well, I can open a dedicated PR for that too.

cpjet64 committed:
Merge branch 'main' into feat/winpath-pathex
